Build futures-util-preview failed with nightly-2019-01-11

OS: ArchLinux
rustup: 1.16.0
cargo: 1.33.0-nightly (34320d212 2019-01-03
rustc: 1.33.0-nightly (d22fa2d87 2019-01-08)

Successfully built with nightly-2019-01-09-x86_64-unknown-linux-gnu.
Failed build with nightly-2019-01-11-x86_64-unknown-linux-gnu.

Output:

```
╭─dcjanus at DCjanusArchLinux in /home/dcjanus/code/futures-rs ((0.3.0-alpha.11) ✔)
╰─λ cargo build                                                                                                                                                                                              22:27:02
   Compiling proc-macro2 v0.4.24
   Compiling unicode-xid v0.1.0
   Compiling libc v0.2.46
   Compiling either v1.5.0
   Compiling rand_core v0.3.0
   Compiling pin-utils v0.1.0-alpha.4
   Compiling slab v0.4.2
   Compiling lazy_static v1.2.0
   Compiling futures-core-preview v0.3.0-alpha.11 (/home/dcjanus/code/futures-rs/futures-core)
   Compiling rand_core v0.2.2
   Compiling futures-channel-preview v0.3.0-alpha.11 (/home/dcjanus/code/futures-rs/futures-channel)
   Compiling futures-sink-preview v0.3.0-alpha.11 (/home/dcjanus/code/futures-rs/futures-sink)
   Compiling iovec v0.1.2
   Compiling rand v0.5.5
   Compiling num_cpus v1.9.0
   Compiling futures-io-preview v0.3.0-alpha.11 (/home/dcjanus/code/futures-rs/futures-io)
   Compiling quote v0.6.10
   Compiling syn v0.15.24
   Compiling proc-macro-hack v0.5.4
   Compiling futures-select-macro-preview v0.3.0-alpha.11 (/home/dcjanus/code/futures-rs/futures-select-macro)
   Compiling futures-util-preview v0.3.0-alpha.11 (/home/dcjanus/code/futures-rs/futures-util)
error[E0308]: mismatched types
   --> futures-util/src/future/maybe_done.rs:110:18
    |
110 |         Pin::set(self, MaybeDone::Done(res));
    |                  ^^^^
    |                  |
    |                  expected mutable reference, found struct `std::pin::Pin`
    |                  help: consider mutably borrowing here: `&mut self`
    |
    = note: expected type `&mut std::pin::Pin<_>`
               found type `std::pin::Pin<&mut future::maybe_done::MaybeDone<Fut>>`

error[E0308]: mismatched types
  --> futures-util/src/future/fuse.rs:53:18
   |
53 |         Pin::set(self.as_mut().future(), None);
   |                  ^^^^^^^^^^^^^^^^^^^^^^
   |                  |
   |                  expected mutable reference, found struct `std::pin::Pin`
   |                  help: consider mutably borrowing here: `&mut self.as_mut().future()`
   |
   = note: expected type `&mut std::pin::Pin<_>`
              found type `std::pin::Pin<&mut std::option::Option<Fut>>`

error[E0308]: mismatched types
  --> futures-util/src/future/into_stream.rs:39:18
   |
39 |         Pin::set(self.as_mut().future(), None);
   |                  ^^^^^^^^^^^^^^^^^^^^^^
   |                  |
   |                  expected mutable reference, found struct `std::pin::Pin`
   |                  help: consider mutably borrowing here: `&mut self.as_mut().future()`
   |
   = note: expected type `&mut std::pin::Pin<_>`
              found type `std::pin::Pin<&mut std::option::Option<Fut>>`

error[E0308]: mismatched types
  --> futures-util/src/try_future/flatten_sink.rs:62:18
   |
62 |         Pin::set(self.as_mut(), FlattenSink(Ready(resolved_stream)));
   |                  ^^^^^^^^^^^^^
   |                  |
   |                  expected mutable reference, found struct `std::pin::Pin`
   |                  help: consider mutably borrowing here: `&mut self.as_mut()`
   |
   = note: expected type `&mut std::pin::Pin<_>`
              found type `std::pin::Pin<&mut try_future::flatten_sink::FlattenSink<Fut, Si>>`

error[E0308]: mismatched types
   --> futures-util/src/try_future/flatten_sink.rs:102:22
    |
102 |             Pin::set(self, FlattenSink(Closed));
    |                      ^^^^
    |                      |
    |                      expected mutable reference, found struct `std::pin::Pin`
    |                      help: consider mutably borrowing here: `&mut self`
    |
    = note: expected type `&mut std::pin::Pin<_>`
               found type `std::pin::Pin<&mut try_future::flatten_sink::FlattenSink<Fut, Si>>`

error[E0308]: mismatched types
  --> futures-util/src/stream/unfold.rs:99:22
   |
99 |             Pin::set(self.as_mut().fut(), Some(fut));
   |                      ^^^^^^^^^^^^^^^^^^^
   |                      |
   |                      expected mutable reference, found struct `std::pin::Pin`
   |                      help: consider mutably borrowing here: `&mut self.as_mut().fut()`
   |
   = note: expected type `&mut std::pin::Pin<_>`
              found type `std::pin::Pin<&mut std::option::Option<Fut>>`

error[E0308]: mismatched types
  --> futures-util/src/try_stream/try_for_each.rs:52:22
   |
52 |             Pin::set(self.as_mut().future(), None);
   |                      ^^^^^^^^^^^^^^^^^^^^^^
   |                      |
   |                      expected mutable reference, found struct `std::pin::Pin`
   |                      help: consider mutably borrowing here: `&mut self.as_mut().future()`
   |
   = note: expected type `&mut std::pin::Pin<_>`
              found type `std::pin::Pin<&mut std::option::Option<Fut>>`

error[E0308]: mismatched types
  --> futures-util/src/try_stream/try_for_each.rs:57:30
   |
57 |                     Pin::set(self.as_mut().future(), Some(future));
   |                              ^^^^^^^^^^^^^^^^^^^^^^
   |                              |
   |                              expected mutable reference, found struct `std::pin::Pin`
   |                              help: consider mutably borrowing here: `&mut self.as_mut().future()`
   |
   = note: expected type `&mut std::pin::Pin<_>`
              found type `std::pin::Pin<&mut std::option::Option<Fut>>`

error[E0308]: mismatched types
   --> futures-util/src/sink/with_flat_map.rs:105:18
    |
105 |         Pin::set(stream, None);
    |                  ^^^^^^
    |                  |
    |                  expected mutable reference, found struct `std::pin::Pin`
    |                  help: consider mutably borrowing here: `&mut stream`
    |
    = note: expected type `&mut std::pin::Pin<_>`
               found type `std::pin::Pin<&mut std::option::Option<St>>`

error: aborting due to 9 previous errors

For more information about this error, try `rustc --explain E0308`.
error: Could not compile `futures-util-preview`.

To learn more, run the command again with --verbose.
```

--- a/.travis.yml
+++ b/.travis.yml
@@ -25,7 +25,7 @@ matrix:
 
     # When updating this, the reminder to update the minimum required version in README.md.
     - name: cargo test (minimum required version)
-      rust: nightly-2018-12-26
+      rust: nightly-2019-01-11
 
     - name: cargo clippy
       rust: nightly

--- a/.travis.yml
+++ b/.travis.yml
@@ -64,8 +64,10 @@ matrix:
       script:
         - cargo build --manifest-path futures/Cargo.toml --features tokio-compat,io-compat
 
+    # Allow build fail until #1396 is fixed.
     - name: cargo build --target=thumbv6m-none-eabi
       rust: nightly
+      env: ALLOW_FAILURES=true
       install:
         - rustup target add thumbv6m-none-eabi
       script:
@@ -99,6 +101,10 @@ matrix:
         - git add "$TRAVIS_TAG" index.html
         - git commit -m "Add API docs for $TRAVIS_TAG"
         - git push origin master
+
+  allow_failures:
+    - rust: nightly
+      env: ALLOW_FAILURES=true
 
 script:
   - cargo test --all --all-features

--- a/README.md
+++ b/README.md
@@ -39,7 +39,7 @@ Now, you can use futures-rs:
 use futures::future::Future; // Note: It's not `futures_preview`
 ```
 
-The current version of futures-rs requires Rust nightly 2018-12-26 or later.
+The current version of futures-rs requires Rust nightly 2019-01-11 or later.
 
 ### Feature `std`
 

--- a/futures-util/src/future/fuse.rs
+++ b/futures-util/src/future/fuse.rs
@@ -50,7 +50,7 @@ impl<Fut: Future> Future for Fuse<Fut> {
             None => return Poll::Pending,
         };
 
-        Pin::set(self.as_mut().future(), None);
+        Pin::set(&mut self.as_mut().future(), None);
         Poll::Ready(v)
     }
 }

--- a/futures-util/src/future/into_stream.rs
+++ b/futures-util/src/future/into_stream.rs
@@ -36,7 +36,7 @@ impl<Fut: Future> Stream for IntoStream<Fut> {
             None => return Poll::Ready(None),
         };
 
-        Pin::set(self.as_mut().future(), None);
+        Pin::set(&mut self.as_mut().future(), None);
         Poll::Ready(Some(v))
     }
 }

--- a/futures-util/src/future/maybe_done.rs
+++ b/futures-util/src/future/maybe_done.rs
@@ -107,7 +107,7 @@ impl<Fut: Future> Future for MaybeDone<Fut> {
                 MaybeDone::Gone => panic!("MaybeDone polled after value taken"),
             }
         };
-        Pin::set(self, MaybeDone::Done(res));
+        Pin::set(&mut self, MaybeDone::Done(res));
         Poll::Ready(())
     }
 }

--- a/futures-util/src/sink/with_flat_map.rs
+++ b/futures-util/src/sink/with_flat_map.rs
@@ -102,7 +102,7 @@ where
                 };
             }
         }
-        Pin::set(stream, None);
+        Pin::set(&mut stream, None);
         Poll::Ready(Ok(()))
     }
 }

--- a/futures-util/src/stream/unfold.rs
+++ b/futures-util/src/stream/unfold.rs
@@ -96,7 +96,7 @@ impl<T, F, Fut, It> Stream for Unfold<T, F, Fut>
     ) -> Poll<Option<It>> {
         if let Some(state) = self.as_mut().state().take() {
             let fut = (self.as_mut().f())(state);
-            Pin::set(self.as_mut().fut(), Some(fut));
+            Pin::set(&mut self.as_mut().fut(), Some(fut));
         }
 
         let step = ready!(self.as_mut().fut().as_pin_mut().unwrap().poll(lw));

--- a/futures-util/src/try_future/flatten_sink.rs
+++ b/futures-util/src/try_future/flatten_sink.rs
@@ -59,7 +59,7 @@ where
             Waiting(f) => try_ready!(f.try_poll(lw)),
             Closed => panic!("poll_ready called after eof"),
         };
-        Pin::set(self.as_mut(), FlattenSink(Ready(resolved_stream)));
+        Pin::set(&mut self.as_mut(), FlattenSink(Ready(resolved_stream)));
         if let Ready(resolved_stream) = self.project_pin() {
             resolved_stream.poll_ready(lw)
         } else {
@@ -99,7 +99,7 @@ where
             Waiting(_) | Closed => Poll::Ready(Ok(())),
         };
         if res.is_ready() {
-            Pin::set(self, FlattenSink(Closed));
+            Pin::set(&mut self, FlattenSink(Closed));
         }
         res
     }

--- a/futures-util/src/try_stream/try_for_each.rs
+++ b/futures-util/src/try_stream/try_for_each.rs
@@ -49,12 +49,12 @@ impl<St, Fut, F> Future for TryForEach<St, Fut, F>
             if let Some(future) = self.as_mut().future().as_pin_mut() {
                 try_ready!(future.try_poll(lw));
             }
-            Pin::set(self.as_mut().future(), None);
+            Pin::set(&mut self.as_mut().future(), None);
 
             match ready!(self.as_mut().stream().try_poll_next(lw)) {
                 Some(Ok(e)) => {
                     let future = (self.as_mut().f())(e);
-                    Pin::set(self.as_mut().future(), Some(future));
+                    Pin::set(&mut self.as_mut().future(), Some(future));
                 }
                 Some(Err(e)) => return Poll::Ready(Err(e)),
                 None => return Poll::Ready(Ok(())),

